internal/pkg/service/buffer/dependencies/definition.go

```diff
@@ -0,0 +1,34 @@
+package dependencies
+
+import (
+	"context"
+
+	definitionRepo "github.com/keboola/keboola-as-code/internal/pkg/service/buffer/definition/repository"
+)
+
+// serviceScope implements ServiceScope interface.
+type definitionScope struct {
+	ServiceScope
+	definitionRepository *definitionRepo.Repository
+}
+
+func (v *definitionScope) DefinitionRepository() *definitionRepo.Repository {
+	return v.definitionRepository
+}
+
+func NewDefinitionScope(ctx context.Context, svcScope ServiceScope) (v DefinitionScope) {
```

Comment:
Manually passing dependencies via a constructor is the recommended way for Go. The problem with the articles is that they show it on too small a scale, so with 3 arguments it looks nice.
- https://reliasoftware.com/blog/best-practices-for-dependency-inversion-in-golang
- https://stackoverflow.com/a/60271405
- https://www.parserdigital.com/blog/2022/05/13/dependency-injection-and-unit-testing-in-golang/

**From the point of view of dependencies, interfaces are syntactic sugar so that you don't have 10 arguments in the constructor and you can easily pass dependencies to an underlying layers.** 

**If you think about it, the reason why you haven't seen it in another language is that Go interfaces are implemented implicitly, without the need to write `implements` keyword. Go interfaces are problematic in other cases (generic programming), here this property - implicit implementation of interfaces - is nice.**

We are using directly arguments passing via a constructor on _lower levels_ (if there is a small number of dependencies), and `dependencies` interface is used in _higher levels,_ together with mocked dependencies containers in test.

Mostly we need to mock Keboola APIs. We use `etcd` as a database everywhere, and there is no problem to create a new namespace for each test. If we were to use a DBAL, we can use e.g. SQLite driver for GORM in tests.

So, what type of system you are programming, what you need to mock, and similar questions affects the decision.

Please see also following documentation:
https://github.com/keboola/keboola-as-code/blob/9f81b1acfbf820f3142e0e6c7a05c540b2358017/internal/pkg/service/common/dependencies/dependencies.go#L1 (it is focussed on CLI)

I would be happy to discuss it in more detail.

Reply:
It is a more complicated solution, without benefits for our use-cases.